### DISCUSSION TITLE
directories not longer supported as bundle contents

### DISCUSSION THREAD

**Discussion by LongHairedHacker** [rauc/rauc#1533] [Q&A] [answered]
Hi,

I just ran into `directories are not supported as bundle contents` when updating rauc in one of our projects.

The [FAQ](https://rauc.readthedocs.io/en/latest/faq.html#why-does-bundle-creation-fail-with-a-not-supported-as-contents-error) says to get in touch when that happens, so here I am.

Let me quickly explain our use case:
We switched to rauc rather late in that projects development cycle, aiming to replacing a custom update solution with something more standard and well-tested.
Due to the switch being as late as it was, we have a lot of legacy systems in the field that require some clean-up once as soon as the first update is installed using rauc.
For that we made extensive use of hooks.

To keep things somewhat organized we split our hooks into multiple files.
The `main-hook` looks like this:
```shell
#!/bin/sh
set -e

for hook in ${RAUC_MOUNT_PREFIX}/bundle/hooks/*.hook; do
	echo "Running hook ${hook}"
	$hook $@ || exit 1
done
```

We have a `hooks` directory with files like: `001-recharge-flux-capacitor.hook` or `002-merge-old-configs.hook`.
The hook directory gets copied to our bundle input directory, along with the `main-hook` file.

Of course, we could just merge all those files into a single large hook, however we also have multiple hardware platforms, that need a slightly different set of hooks for each. 
So for each bundle the hooks directory is populated from a directory of common hook scripts and another directory with platform specific ones.

Is there a specific reason not to support directories as bundle content, other than being unable to use hardlinks for `.rauc-workdir`?
 If not, rauc could just recursively recreate the directory structure in `.rauc-workdir` and hardlink the individual files over.
(I'd be happy to provide a pull-request for that.)

**Comment by ejoerns** [comment]
> Is there a specific reason not to support directories as bundle content, other than being unable to use hardlinks for .rauc-workdir?
If not, rauc could just recursively recreate the directory structure in .rauc-workdir and hardlink the individual files over.
(I'd be happy to provide a pull-request for that.)

Yes, the reason is that it would require explicit handling and code and that so far we have not seen a real use case for having directories in the bundle.

From my perspective, for organizing hooks directories might not be strictly necessary but *could* be helpful, yes.

I'd prefer an explicit exception instead of recursively adding all directories, but I have no good clue what this could look like.
Only options I see are

* allowing `hooks/` only
* introducing a setting in the manifest
* doing it as you suggested and (recursively) adding all directories

Maybe @jluebbe has an opinion on this.

For now, we have at least re-allowed using local symlinks in https://github.com/rauc/rauc/pull/1538 since this just works as-is.

**Reply by ejoerns** [reply, marked as answer]
Support for (non-hidden) directories in bundles was re-added in RAUC 1.15 (https://github.com/rauc/rauc/pull/1625).
